### PR TITLE
improving docs for v0.3 + fix markdown backend

### LINKED ISSUE #85: document 0.3 changes and finalize release
**edit**: consolidated a few comments with todo
**edit 2**: changed my mind again, release of 0.3 will require only to document past changes in order to be completed

having completed the breaking change:
- [x] make `codeAsInSource` the default and `codeFromAst` an option, see #94 

the only really required thing in order to release 0.3 is:

- [ ] document new stuff (and also stuff from 0.2 minor versions), see #81 and recent PRs
  - [ ] added a new newNbSlimBlock and newNbCodeBlock
  - [ ] nbText now does not read code
  - [ ] new nbTextWithCode that does read code
  - [ ] config utils for a theme such as nimibook
  - [ ] trim readme and it will NOT to depend on index.nim
  - [ ] add to readme a deploy section which documents how to deploy with GitHub pages
  - [ ] make links in docs relative so that they work in deploy preview (already done in js/karax pr, see)
  - [ ] nbRawOutput

all this other stuff can be safely added in 0.3.x updates, I will create issues for each one of them and grouped them in a project/milestone.
- [ ] put back highlighting js #68
- [ ] put back untyped nbFile (see discussion in #34)
- [ ] add test for nbFile
- [ ] add test for nbImage
- [ ] add also nbAudio and nbVideo (with tests) and a media example
- [ ] add nbShell (wraps osproc.execCmdEx) (completes #34)
- [ ] add a document that shows how to use `nbPython` and maybe offers `nbPythonCode` implementation (see [this commit](https://github.com/pietroppeter/nimib/pull/83/commits/fedd4c68baab1c849505c3d5ec455af17c77dac9)) as a novelty (and also as a way to show how to do a custom code block)
- [ ] rethink/review markdown related stuff (both for markdown backend and for converting to markdown) and in general change the documentation:
  - [ ] readme.md will not be generated anymore from nim (too clumsy to fix a small typo in readme) - instead index.nim might depend on readme (partial inclusion).
  - [ ] add a md2html.nim file that converts a given md file to html using default theme (as it happens in nimibook)
  - [ ] this same md2html file will also be saved with markdown backend
  as md2html.md
  - [ ] and md2html.md will be converted itself as a md2html2md2html.html (which might be slightly different from md2html)
- [ ] config file found should report path relative to initDir
- [ ] NbBlock as a container: I think this can be done without breaking changes and it opens up many possibilities of use cases.

### PR DESCRIPTION
**Pull request by pietroppeter**
fixes #85 (last step before tagging and releasing)

- improve changelog
- make sure all important changes (recent and older) are documented

details:

- [x] improving changelog and integrating thanks in changelog
  - [x] 0.1
  - [x] 0.1.x
  - [x] 0.2
  - [x] 0.2.x
  - [x] 0.3
  - [x] move changelog to separate file
- [x] improve documentation for some of the recent changes (from 0.2.0 until now)
  - [x] stuff in 0.2.0 that went un(der)documented
    - [x] add list of command line options generated with `nbInit`
  - [x] changes in 0.2.x
    - [x] 0.2.2: nbFile
  - [x] other 0.3 improvements by Hugo #80
    - [x] nbRawOutput
    - [x] nbClearOutput
  - [x] release 0.3 stuff #81
    - [x] newNbCodeBlock
    - [x] newNbSlimBlock
    - [x] new nbTextWithCode that does read code
    - [x] example files.nim for File
  - [x] nbPython # 83
    - [x] nbInitPython and nbPython
  - [x] nim to javascript #88 - added a section
    - [x] nbCodeToJs (or by pieces nbInitToJs, addCodeToJs, addToDocAsJs)
    - [x] nbKaraxCode
  - [x] CodeAsInSource now the default:
    - [x] remove bullet point in API section
    - [x] add section "Code Capture"
  - [x] make sure to rerun nimble readme to have readme updated (currently missing the interactivity docs)
- [x] other changes
  - [x] turned off warning for unused imports in `nimib.nim` (#103)
  - [x] fix md output of index.nim (bug in markdown backend?)

- some other improvements that could be done later: https://github.com/pietroppeter/nimib/issues/110

### CODE REVIEW

**Comment by pietroppeter** [comment]
ok, this is done. After this is merged I would tag and release.
@HugoGranstrom would you give it a look?

some small details:
- I did not document `nbClearOutput` because I cannot remember what was its use case.
- I removed random from import (I think it was used in js code generation but now it should not be anymore needed)
- I am not sure if it is worth adding at the end of nimib.nim a pragma to turn off unusedImport warning (a normal run of index.nim would still mark sequtils and strutils as not used). I also found very little documentation of this, basically only https://forum.nim-lang.org/t/5461#34187

**Comment by HugoGranstrom** [comment]
> ok, this is done. After this is merged I would tag and release.
> @HugoGranstrom would you give it a look?

:confetti_ball: 
I'll take a closer look later tonight but I'll answer your questions:

> I did not document nbClearOutput because I cannot remember what was its use case.

It was mainly for nimiSlides to save some screen estate if the output of the code block wasn't important. (ggplotnim for example echo:ed some really noisy stuff for example)

> I removed random from import (I think it was used in js code generation but now it should not be anymore needed)

Ah yes forgot about that one. Also I saw that they have released a new Karax release so we should be able to drop the `@#head` in the `nimib.nimble` file.

> I am not sure if it is worth adding at the end of nimib.nim a pragma to turn off unusedImport warning (a normal run of index.nim would still mark sequtils and strutils as not used). I also found very little documentation of this, basically only https://forum.nim-lang.org/t/5461#34187

As long as it doesn't affect the user's imports, I think that would be a welcome addition :+1:

**Comment by pietroppeter** [comment]
> > I did not document nbClearOutput because I cannot remember what was its use case.
> 
> It was mainly for nimiSlides to save some screen estate if the output of the code block wasn't important. (ggplotnim for example echo:ed some really noisy stuff for example)

ok, makes sense. added explanation.

> > I removed random from import (I think it was used in js code generation but now it should not be anymore needed)
> 
> Ah yes forgot about that one. Also I saw that they have released a new Karax release so we should be able to drop the `@#head` in the `nimib.nimble` file.

also update Karax version

> > I am not sure if it is worth adding at the end of nimib.nim a pragma to turn off unusedImport warning (a normal run of index.nim would still mark sequtils and strutils as not used). I also found very little documentation of this, basically only https://forum.nim-lang.org/t/5461#34187
> 
> As long as it doesn't affect the user's imports, I think that would be a welcome addition 👍

good idea. I checked and indeed it does not affect user's imports. Apparently it only affects import of the module where it is added. So added.

Instead I am having issues with the markdown backend. I did not realize it is actually using parts of html backend and not sure why... will look into that and fix it.

**Review by HugoGranstrom — CHANGES_REQUESTED** [review]
Nothing major really, just a few nitpicks. It looks very good :rocket: 

But wow! There has been so much happening to nimib over the last year, I can't believe it was just over a year since v0.1 was released :exploding_head:

File: docsrc/index.nim [hunk 1]
```diff
@@ -123,6 +127,41 @@ Customization over the default is mostly achieved through nim-mustache or changi
...
+
+### creating custom blocks
+
+* `newNbCodeBlock(cmd: string, body, blockImpl: untyped)`: template that can be used to create custom
+  code block that will need both a `body` and an implementation which might make use of `body`.
+  Also, the source code in `body` is read.
+  Example blocks created with `newNbCodeBlock` are `nbCode` and `nbTextWithCode`.
+* `newNbSlimBlock(cmd: string, blockImpl: untyped)`: template that can be used to create
+  a custom block that does not need a separate `body`.
+  Example blocks created with `newNbSlimBlock` are `nbText`, `nbImage` and `nbFile`.
```
[the remark this answers is not in the document]
Comment (pietroppeter) [reply docsrc/index.nim:151]: you are right, will remove the ref to nbFile

File: changelog.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+* new template `nbRawOutput` that renders raw html (#80)
+* new command `nbClearOutput` that removes output from last block processed (#80)
+* new templates for creating code blocks `newNbCodeBlock` and `newNbSlimBlock`
+  for creating custom blocks and related changes (#81):
+  - `newNbCodeBlock`: captures source code of `body`
+  - `newNbSlimBlock`: block without a `body` (and in particular no capture of source)
+  - all nimib "native" blocks now use one of the two mechanism above
+  - now `nbText` does NOT contain its source
+  - new `nbTextWithCode` that does contain code source
+* untyped version of `nbFile` removed, new example document `files.nim`, changed the rendering of `nbFile`
```
[the remark this answers is not in the document]
Comment (pietroppeter) [reply changelog.md:71]: indeed I do not know it it was removed at some point or not (I already forgot if I added it back in some previous PR)

**Comment by pietroppeter** [comment]
> But wow! There has been so much happening to nimib over the last year, I can't believe it was just over a year since v0.1 was released 🤯

indeed when writing the history/changelog I was also surprised by how many changes there were!

fixed your remarks and also fixed the markdown backend which required some non trivial changes. Also updated the rendering documentation which was indeed obsolete.

**Review by HugoGranstrom — APPROVED** [review]

**Comment by HugoGranstrom** [comment]
> fixed your remarks and also fixed the markdown backend which required some non trivial changes. Also updated the rendering documentation which was indeed obsolete.

:rocket: :partying_face: Awesome! It felt very typical that it was the thing we will refactor next that was misbehaving, I'll take that as a sign :rofl: 

So all that's left now is to go through the preview one more time and then release 0.3? :exploding_head:

**Comment by pietroppeter** [comment]
> So all that's left now is to go through the preview one more time and then release 0.3? 🤯

Yep! I will also do another review later on

File: docsrc/index.nim [hunk 3]
```diff
@@ -1,17 +1,20 @@
...
+when defined(mdOutput):
+  echo "using Markdown backend"
+  nbInitMd
+else:
+  nbInit
 nb.title = "Nimib Docs"
 
 let
   repo = "https://github.com/pietroppeter/nimib"
   docs = if defined(useMdBackend): "https://pietroppeter.github.io/nimib" else: "."
```
Comment (HugoGranstrom) [inline docsrc/index.nim:14]:
```suggestion
  docs = if defined(mdOutput): "https://pietroppeter.github.io/nimib" else: "."
```
The readme links were incorrect

**Comment by pietroppeter** [comment]
made another pass but seems good to go! 🥳

**Comment by pietroppeter** [comment]
you are right, will do

**Pull request merged**